File: src/Other/MicroPython/composite.py
class FileLeaf:
    def __init__(self, size):
        self._size = size

    def size(self):
        return self._size


class FolderComposite:
    def __init__(self, children):
        self._children = list(children)

    def size(self):
        total = 0
        for child in self._children:
            total += child.size()
        return total


readme = FileLeaf(2)
docs = FolderComposite((FileLeaf(3), FileLeaf(5)))
root = FolderComposite((readme, docs))

print("leaf={}".format(readme.size()))
print("docs={}".format(docs.size()))
print("root={}".format(root.size()))

File: src/Web/TypeScriptTS/composite.ts
interface Component {
  size(): number;
}

class FileLeaf implements Component {
  constructor(private readonly bytes: number) {}
  size(): number { return this.bytes; }
}

class FolderComposite implements Component {
  constructor(private readonly children: Component[]) {}
  size(): number { return this.children.reduce((total, child) => total + child.size(), 0); }
}

const readme: Component = new FileLeaf(2);
const docs: Component = new FolderComposite([new FileLeaf(3), new FileLeaf(5)]);
const root: Component = new FolderComposite([readme, docs]);

console.log(`leaf=${readme.size()}`);
console.log(`docs=${docs.size()}`);
console.log(`root=${root.size()}`);
